### PR TITLE
Add methods for publishing in chunks.

### PR DESCRIPTION
**Pull request by embhorn**
One highly requested feature has been a method for publishing in chunks, i.e. use a buffer that is smaller than the total payload. This update adds two methods: A callback that is executed from the new ```MqttClient_Publish_ex``` API; and it is now possible to simply call ```MqttClient_Publish``` successively when the payload buffer size is smaller than the total payload size as set in the ```MqttPublish``` structure parameter. The example for the client firmware push was modified to use the new publish callback.

Additionally a bug in the MQTTv5 assigned client ID feature was fixed, along with a defect in the ```MQTTCtx``` structure.

### CODE REVIEW

File: examples/firmware/fwpush.c [hunk 1]
```diff
@@ -388,40 +423,51 @@ int fwpush_test(MQTTCtx *mqttCtx)
...
             FALL_THROUGH;
         }
 
         case WMQ_PUB:
         {
             mqttCtx->stat = WMQ_PUB;
 
-            /* Publish Topic */
-            rc = MqttClient_Publish(&mqttCtx->client, &mqttCtx->publish);
+            /* setup publish message */
```
Comment (dgarske) [inline examples/firmware/fwpush.c]: This doesn't work correctly with non-blocking mode enabled. We setup the public message in the previous state so this WMQ_PUB can be called again in the MQTT_CODE_CONTINUE case.

Example:

```
./configure --enable-nonblock --enable-debug --disable-shared && make
./examples/firmware/fwpush


MQTT Firmware Push Client: QoS 2, Use TLS 0
MQTT Net Init: Success (0)
MQTT Init: Success (0)
MQTT Socket Connect: Success (0)
MQTT Connect: Success (0)
MQTT Connect Ack: Return Code 0, Session Present 0
Firmware File README.md is 17251 bytes
Firmware Message: Sig 74 bytes, Key 65 bytes, File 17251 bytes
Firmware File README.md is 17251 bytes
Firmware Message: Sig 74 bytes, Key 65 bytes, File 17251 bytes
MQTT Publish: Topic wolfMQTT/example/firmware, Unknown (-13)
MQTT Socket Disconnect: Success (0)
```

Normal blocking mode reports:

```
MQTT Firmware Push Client: QoS 2, Use TLS 0
MQTT Net Init: Success (0)
MQTT Init: Success (0)
MQTT Socket Connect: Success (0)
MQTT Connect: Success (0)
MQTT Connect Ack: Return Code 0, Session Present 0
Firmware File README.md is 17251 bytes
Firmware Message: Sig 74 bytes, Key 65 bytes, File 17251 bytes
MQTT Publish: Topic wolfMQTT/example/firmware, Success (0)
MQTT Disconnect: Success (0)
MQTT Socket Disconnect: Success (0)
```

File: wolfmqtt/mqtt_types.h [hunk 2]
```diff
@@ -105,6 +105,7 @@ enum MqttPacketResponseCodes {
     MQTT_CODE_ERROR_STAT = -10,
     MQTT_CODE_ERROR_PROPERTY = -11,
     MQTT_CODE_ERROR_SERVER_PROP = -12,
+    MQTT_CODE_ERROR_CALLBACK = -13,
```
Comment (dgarske) [inline wolfmqtt/mqtt_types.h]: Please add this to `MqttClient_ReturnCodeToString`.

File: examples/firmware/fwpush.c [hunk 3]
```diff
@@ -388,40 +423,51 @@ int fwpush_test(MQTTCtx *mqttCtx)
...
+            XMEMSET(&mqttCtx->publish, 0, sizeof(MqttPublish));
+            mqttCtx->publish.retain = mqttCtx->retain;
+            mqttCtx->publish.qos = mqttCtx->qos;
+            mqttCtx->publish.duplicate = 0;
+            mqttCtx->publish.topic_name = mqttCtx->topic_name;
+            mqttCtx->publish.packet_id = mqtt_get_packetid();
+            mqttCtx->publish.buffer_len = FIRMWARE_MAX_BUFFER;
+            mqttCtx->publish.buffer = WOLFMQTT_MALLOC(FIRMWARE_MAX_BUFFER);
+
```
Comment (dgarske) [inline examples/firmware/fwpush.c]: For example/reference can we add setting `mqttCtx->publish.ctx` for user context and show using it in the `mqtt_publish_cb` function?

File: examples/firmware/fwpush.c [hunk 4]
```diff
@@ -75,6 +75,52 @@ static int mqtt_message_cb(MqttClient *client, MqttMessage *msg,
...
+        bytes_read = fread(publish->buffer, 1, publish->buffer_len, fp);
+        if (bytes_read != 0) {
+            ret = (int)bytes_read;
+        }
+    }
+    if (feof(fp)) {
+        fclose(fp);
+    }
+#endif
+    return(ret);
```
Comment (dgarske) [inline examples/firmware/fwpush.c]: For consistency with wolf code standards can we use `return ret;` here?

File: examples/firmware/fwpush.c (hunk 1 above)
Comment (embhorn) [reply examples/firmware/fwpush.c]: Check

File: wolfmqtt/mqtt_types.h (hunk 2 above)
Comment (embhorn) [reply wolfmqtt/mqtt_types.h]: Check

File: examples/firmware/fwpush.c (hunk 3 above)
Comment (embhorn) [reply examples/firmware/fwpush.c]: The fwpush app currently stores the computed ```FirmwareHeader``` buffer in ```mqttCtx->publish.ctx```. Then in the callback the first publish buffer is created using the stored data.

Did you have something else in mind?

File: examples/firmware/fwpush.c (hunk 4 above)
Comment (embhorn) [reply examples/firmware/fwpush.c]: Check

File: examples/firmware/fwpush.c (hunk 3 above)
Comment (dgarske) [reply examples/firmware/fwpush.c]: If that `mqttCtx->publish.ctx` is already used then I suggest adding an additional parameter to `MqttClient_Publish_ex` to allow the user to supply a custom void* context to the callback. Modify the `MqttPublishCb` typedef to also pass this void* in.

**Pull request merged**